### PR TITLE
Use I;16 mode for 9-bit JPEG 2000 images

### LINKED ISSUE #7899: JPEG2000: 9-bit images use L mode instead of I;16
### What did you do?
9-bit precision JPEG 2000 codestream with a .txt extension: [9bit.txt](https://github.com/python-pillow/Pillow/files/14734392/9bit.txt) (rename to 9bit.j2k)
```python
with Image.open("9bit.j2k") as im:
    assert im.mode == "I;16"
    assert im.size == (128, 128)
```

### What did you expect to happen?
No exceptions

### What actually happened?
`AssertionError: assert 'L' == 'I;16'`

### What are your OS, Python and Pillow versions?

* OS: Ubuntu 22.04
* Python: 3.12
* Pillow: Current `main` 19cd94bdb3699abe11d5e437406daa66f7f5cc17

### PR DESCRIPTION
**Pull request by scaramallion**
Fixes #7899 .

Changes proposed in this pull request:

 * Fixes the mode setting for 9-bit J2K images
 * Adds a 9-bit test image 

Reference: [ISO 15444-1](https://www.itu.int/rec/T-REC-T.800-201511-S/en), Table A-11 

image

### CODE REVIEW

**Comment by scaramallion** [comment]
Now that I look at it again, why is it using `yrsiz[0]` to determine the mode? `yrsiz` is the vertical separation of the component. I would've though it'd be `(ssiz[0] & 0x7F) + 1` for signed images and just `ssiz[0] + 1` for unsigned.

**Comment by scaramallion** [comment]
Oh, there's another bug further up in parsing the rest of the header that means the `ssiz` value is actually in `yrsiz`.

This should only be unpacking from [39:41] instead of from [36:39].
```python
for i in range(csiz):
        ssiz[i], xrsiz[i], yrsiz[i] = struct.unpack_from(">BBB", siz, 36 + 3 * i)
```

**Comment by scaramallion** [comment]
Ah, there's another bug further up in parsing the header that's causing the `ssiz` value to end up in `yrsiz`, this should be `38 + 3 * i`:

```python
for i in range(csiz):
        ssiz[i], xrsiz[i], yrsiz[i] = struct.unpack_from(">BBB", siz, 36 + 3 * i)
```

**Comment by scaramallion** [comment]
Should I fix that as well, or should it be a separate issue?

**Comment by radarhere** [comment]
I think either should be fine.
If it's something producing a problem that you can test for, I'd say create a new PR with a test, otherwise here is good.

**Review by radarhere — APPROVED** [review]

File: Tests/test_file_jpeg2k.py [hunk 1]
```diff
@@ -446,3 +446,10 @@ def test_plt_marker() -> None:
         hdr = out.read(2)
         length = _binary.i16be(hdr)
         out.seek(length - 2, os.SEEK_CUR)
+
+
+def test_9bit():
+    # Test 9-bit images are decoded to I;16
```
Comment (radarhere) [inline Tests/test_file_jpeg2k.py:452]:
```suggestion
```
I think this is fairly obvious from the rest of the test

**Pull request merged**